rosbag2_transport/src/rosbag2_transport/recorder.cpp

```diff
@@ -395,18 +395,29 @@ const rosbag2_cpp::Writer & RecorderImpl::get_writer_handle()
 
 void RecorderImpl::pause()
 {
-  paused_.store(true);
-  RCLCPP_INFO_STREAM(node->get_logger(), "Pausing recording.");
+  std::lock_guard<std::recursive_mutex> state_lock(state_transition_mutex_);
+  if (paused_) {
+    RCLCPP_DEBUG(node->get_logger(), "Recorder is alreadyin pause state.");
+  } else {
+    paused_.store(true);
+    RCLCPP_INFO_STREAM(node->get_logger(), "Pausing recording.");
+  }
 }
 
 void RecorderImpl::resume()
 {
-  paused_.store(false);
-  RCLCPP_INFO_STREAM(node->get_logger(), "Resuming recording.");
+  std::lock_guard<std::recursive_mutex> state_lock(state_transition_mutex_);
+  if (!paused_) {
+    RCLCPP_DEBUG(node->get_logger(), "Already in the recording.");
+  } else {
+    paused_.store(false);
+    RCLCPP_INFO_STREAM(node->get_logger(), "Resuming recording.");
+  }
 }
 
 void RecorderImpl::toggle_paused()
 {
+  std::lock_guard<std::recursive_mutex> state_lock(state_transition_mutex_);
```

Comment:
> a concern about using recursive_mutex for protecting state transition.

`resume` or `pause` would be called via another thread, so without having the mutex lock in them recursively, the behavior is not gonna be transactional. for example, this thread comes in here with `state_transition_mutex_` and check it is paused, then another thread call `resume` without `state_transition_mutex_`, then this `resume` does print `Already in the recording.`.

if this is acceptable, we can remove recursive mutex lock, i think. what do you think? @MichaelOrlov

Reply:
@fujitatomoya I don't think that we need to remove the mutex lock at all in resume and toggle. But rather chnage it to the non-recursive mutex.

I will do a prototype.

Reply:
@fujitatomoya See my new commit https://github.com/ros2/rosbag2/pull/1547/commits/4fba7a49ebe3993a6c0d2e86deb0f41ee8957166 on your branch